### DISCUSSION TITLE
Is Appender the fastest API to import data?

### DISCUSSION THREAD

**Discussion by wangfenjin** [duckdb/duckdb#2085] [General]
Recently there is an article [Towards Inserting One Billion Rows in SQLite Under A Minute](https://avi.im/blag/2021/fast-sqlite-inserts/), I test it in my local machine MacBook Pro (16-inch, 2019):

```shell
# run the fastest file in the article, insert 100M rows
# https://github.com/avinassh/fast-sqlite3-inserts/blob/master/src/bin/basic_batched.rs
time cargo run --bin basic_batched
# output
# cargo run --bin basic_batched  193.52s user 5.81s system 98% cpu 3:22.47 total
```

It takes about 193s to insert 100M rows. Then I test insert 10M using duckdb-rs:
```shell
# https://github.com/wangfenjin/duckdb-rs/blob/main/examples/appender.rs
# In github code is 1M, but I change to 10M when run this test
time cargo run --example appender
# output
# cargo run --example appender  286.56s user 8.84s system 97% cpu 5:01.49 total
```

Which means when insert data, DuckDB is 10 times slower than SQLite.

Not sure if it's the right way to do this?

**Comment by Mytherin** [comment]
The appender is not particularly efficient because it performs a commit every 1024 rows. It would not surprise me if that causes a >10x performance hit. This needs to be reworked, either by increasing the caching size or keeping a transaction open for the duration of the appender. I can have a look at that next week.

**Reply by Mytherin** [reply]
Actually this is only true if you run the appender in auto commit mode, if you wrap the appender calls in Begin Transaction and Commit you should get better performance.

Still not great because the binding happens every 1024 rows, but it should improve things.

**Comment by Mytherin** [comment]
I made a PR that should fix these issues here: #2087

After that PR I can insert 100M rows in 18s without a primary key, and 80s with a primary key, at least using the C++ API.

I suspect there might still be some performance problems remaining within the Rust layer which you might want to have a look at and profile/fix.

**Reply by wangfenjin** [reply]
Great, I’ll try!

**Comment by mskyttner** [comment]
I tried sampling 100M rows conforming to the schema in the article using R. This step took about 9 seconds on my laptop and did not involve duckdb and might therefore not be so interesting to benchmark?

With the sample data then available in memory, that data can be "registered" with duckdb. On this step, I ran into an issue that should be resolved by https://github.com/duckdb/duckdb/pull/2044 but not in my installed version of duckdb, it seems (I'm using v 0.2.7).

Making a query against this data using in-memory duckdb (`select * from mydata where age = 5 and active = 1 limit 1;`) took about 0.7 seconds. 

To get a "persisted database" file, I tried to persist the in-memory data to disk using the parquet format with snappy compression (`COPY mydata TO 'mydata.parquet' (FORMAT 'PARQUET', CODEC 'SNAPPY')`). This took about 26 seconds. 

Querying that .parquet file using the same (initial) query above took 0.05 second, faster than above probably due to optimizations described [here](https://duckdb.org/2021/06/25/querying-parquet.html).

The R code is [available as a gist](https://gist.github.com/mskyttner/fe8f425129eabb6575ca321c504bde9c).

So now I wonder, for the original question in this issue, ie "Is the Appender the fastest API to import data?", does the COPY statement use the Appender under the hood? If so, does it do periodic commit transactions?
